python/pyspark/sql/pandas/serializers.py

```diff
@@ -728,8 +743,9 @@ def load_stream(self, stream):
             dataframes_in_group = read_int(stream)
 
             if dataframes_in_group == 2:
+                # We need to fully load the left batches, but we can lazily load the right batches
```

Comment:
It still may cause memory issue, is it possible to do it in this way

```
cogroup: (Iterator[InternalRow], Iterator[InternalRow])

(left, right) = cogroup

while (left.hasNext || right.hasNext) {
   if (left.hasNext) {
      writeBatch(... , "left")
   } else {
      writeBatch(A empty Batch , "left")
   }

   if (right.hasNext) {
      writeBatch(... , "right")
   } else {
      writeBatch(A empty Batch , "right")
   }
}

```

Reply:
The problem is more on the Python side, there's no way to force the user to fully use the left side before using the right side that I can think of.

Reply:
I suggest we narrow this PR down to `Grouped.ApplyInArrow`.
IMHO, `CoGrouped` is rarely used, and it is not very clear how to generate the records.

Reply:
The JVM side for grouped map was adapted from and consolidate with the cogroup code to use an Arrow stream per group, and that has benefits of at least not collecting an entire group on the JVM side before serializing it, so I think that should stay. I can undo the Python side changes and leave only the Table -> Table API for cogroup if that's desired. There are at least _some_ improvements to this implementation of cogroup, especially with being able to return an iterator of RecordBatches

Reply:
I hope we can address the memory issue in `applyInArrow`, since it is not completely resolved for `CoGrouped.applyInArrow`, I'd rather let it alone.
WDYT? @HyukjinKwon @xinrong-meng

Reply:
We should address memory issue, yes. that's why we allow iterator of record batches?

Reply:
Sorry the main question is, do we update the `Cogrouped.applyInArrow` in this PR as well, which is _better_ than it currently is but still not completely lazy (the left side has to be eagerly loaded in memory on the Python side), or do we not touch the Cogrouped API since it's not completely optimal and address that later?

Reply:
If we're going to address them in a separate PR, I am fine. Just wanted to make sure we fix it.

Reply:
Okay I'll take out the Cogroup Python changes and make a separate PR for them